src/hotspot/share/memory/metaspace.hpp

```diff
@@ -113,8 +117,32 @@ class Metaspace : public AllStatic {
   static MetaWord* allocate(ClassLoaderData* loader_data, size_t word_size,
                             MetaspaceObj::Type type);
 
-  static bool contains(const void* ptr);
-  static bool contains_non_shared(const void* ptr);
+  // Returns true if the pointer points into class space, non-class metaspace, or the
+  // Klass portion of the CDS archive.
+  static bool contains(const void* ptr) {
+    return is_in_shared_metaspace(ptr) || // in cds
+           is_in_class_space(ptr) ||      // in class space
+           is_in_nonclass_metaspace(ptr); // in one of the non-class regions?
+  }
+
+  // Returns true if the pointer points into class space or into non-class metaspace
+  static bool contains_non_shared(const void* ptr) {
+    return is_in_class_space(ptr) ||      // in class space
+           is_in_nonclass_metaspace(ptr); // in one of the non-class regions?
+  }
+
+  // Returns true if pointer points into the CDS klass region.
+  static bool is_in_shared_metaspace(const void* ptr);
+
+  // Returns true if pointer points into one of the non-class-space metaspace regions.
+  static bool is_in_nonclass_metaspace(const void* ptr);
+
+  // Returns true if pointer points into class space, false if it doesn't or if
+  // there is no class space. Class space is a contiguous region, which is why
+  // two address comparisons are enough.
+  static inline bool is_in_class_space(const void* ptr) {
+    return ptr < _class_space_end && ptr >= _class_space_start;
+  }
```

Comment:
So this does _not_ include the Klasses in the CDS archive, it really is only useful for deallocation.  I don't need this to be a comment update, just making sure I follow.